newsrevenuehub/rev-engine
# DEV-2296: Fix CSP errors with rich text editor

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

#### What's this PR do?

Fixes CSP problems with our rich text editor. It sets text alignment and font size using inline styles (e.g. `<p style="text-align:center">`) which our CSP policy blocks. The approach here is to convert these styles to CSS classes which we can style accordingly.

Implementation details:

- Moved the rich text editor to a standalone component since it seems like we'll probably use it outside of the rich text editor block one day.
- Added a new component to display the content created in a rich text editor, applying styles to the CSS classes set by the editor.
- Replaced the `draftjs-to-html` package with `draft-js-export-html` because it didn't have the flexibility needed to replace inline styles with classes.
- Migrated the DRichText component to TypeScript.

#### Why are we doing this? How does it help us?

Gets us closer to being able to enforce CSP in production.

#### How should this be manually tested? Please include detailed step-by-step instructions.

Testing the code change:

1. Edit a contribution page and put a rich text editor block in the page content if it isn't there already.
2. Put every possible style into the block :-/
  a. Bold
  b. Italics
  c. Underline
  d. Strikethrough
  e. H2
  f. H3
  g. Blockquote
  h. All font size variations :-/: 13, 14, 15,16, 24
  i. Bulleted list (with indent)
  j. Numbered list (with indent)
  k. Left-aligned text
  l. Center-aligned text
  m. Right-aligned text
  n. Links
3. Save changes to the block.
4. Confirm the page preview reflects all style changes correctly.
5. Save changes to the page.
6. Confirm the page as published reflects all these styles too.

🚨 This PR potentially changes how existing pages will display (but shouldn't). I feel pretty strongly we should deploy this to staging and review all our existing contribution pages to make sure they look right.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

n/a

#### Are there any smells or added technical debt to note?

No.

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-2296

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

Although this fixes content going forward, this will not fix CSP violations for existing content. I created https://news-revenue-hub.atlassian.net/browse/DEV-3033 to capture that work.
